### PR TITLE
Add native monocular 3D detection support

### PR DESCRIPTION
**Pull request by Dorablank**
## Summary

- Add a native `detect3d` task to the existing YOLO Python and CLI surfaces: `train`, `val`, `predict`, and `export`.
- Add lightweight YOLO11 and YOLO26 monocular 3D heads with box-relative center, direct depth, dimensions, MultiBin orientation, projected-corner supervision, and localization-quality scoring.
- Keep camera calibration synchronized through letterbox, horizontal flip, and multi-scale preprocessing, and expose decoded 3D geometry through `Results` and ONNX output.
- Add generic 3D metrics, KITTI R40 evaluation, projected 3D visualization, task-specific plots, documentation, and focused regression coverage.

## Engineering value

This PR is intended as an engineering integration rather than a SOTA claim. It provides a complete, lightweight monocular 3D workflow inside the normal Ultralytics task architecture instead of requiring a separate training/evaluation project. Existing tasks retain their defaults and checkpoint-selection behavior.

## KITTI results

The primary results use the standard Chen split (3,712 train / 3,769 val images). KITTI R40 AP3D uses the class-standard IoU thresholds.

| Model | Params | GFLOPs | Car mod. | Pedestrian mod. | Cyclist mod. | Mean mod. | PT latency | ONNX CUDA latency |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| YOLO26n-3d, seed 0 | 3.12M | 8.1 | 9.67 | 4.65 | 5.37 | 6.56 | 5.925 ms | 2.765 ms |
| YOLO26s-3d, seed 1 | 10.73M | 25.1 | 13.18 | 6.06 | 5.10 | 8.11 | 6.694 ms | 3.017 ms |

<details>
<summary>Full KITTI R40 tables and supplementary 80/20 results</summary>

Values are `Easy / Moderate / Hard` percentages.

### Chen split (primary)

| Model | Class | AP3D | APBEV | AOS |
|---|---|---|---|---|
| YOLO26n-3d, seed 0 | Car | 12.87 / 9.67 / 8.13 | 21.18 / 15.76 / 13.58 | 94.23 / 86.18 / 79.73 |
| YOLO26n-3d, seed 0 | Pedestrian | 5.84 / 4.65 / 3.69 | 6.79 / 5.52 / 4.51 | 50.26 / 42.76 / 37.91 |
| YOLO26n-3d, seed 0 | Cyclist | 10.40 / 5.37 / 4.92 | 10.88 / 5.61 / 5.37 | 41.89 / 27.25 / 26.11 |
| YOLO26s-3d, seed 1 | Car | 17.95 / 13.18 / 11.21 | 26.30 / 19.66 / 17.03 | 96.40 / 90.61 / 82.69 |
| YOLO26s-3d, seed 1 | Pedestrian | 8.28 / 6.06 / 4.91 | 9.96 / 7.25 / 5.99 | 56.33 / 47.84 / 41.46 |
| YOLO26s-3d, seed 1 | Cyclist | 10.31 / 5.10 / 4.96 | 10.96 / 5.72 / 5.39 | 47.60 / 31.24 / 30.03 |

### Deterministic stratified 80/20 split (supplementary)

These results use 5,985 train / 1,496 val images. They must not be compared directly with the Chen split or external KITTI results because correlated frames can cross the image-level split.

| Model | Class | AP3D | APBEV | AOS |
|---|---|---|---|---|
| YOLO26n-3d, seed 2 | Car | 65.57 / 51.89 / 46.44 | 74.00 / 58.89 / 53.25 | 98.60 / 98.03 / 95.36 |
| YOLO26n-3d, seed 2 | Pedestrian | 24.60 / 19.84 / 17.69 | 27.09 / 22.01 / 19.80 | 89.25 / 84.34 / 81.18 |
| YOLO26n-3d, seed 2 | Cyclist | 35.11 / 28.54 / 27.68 | 38.47 / 30.94 / 30.20 | 89.20 / 87.40 / 85.09 |
| YOLO26s-3d, seed 0 | Car | 75.05 / 60.84 / 54.68 | 81.15 / 67.45 / 61.28 | 99.18 / 99.06 / 96.52 |
| YOLO26s-3d, seed 0 | Pedestrian | 30.53 / 24.46 / 21.13 | 32.19 / 25.77 / 22.34 | 94.11 / 88.52 / 85.19 |
| YOLO26s-3d, seed 0 | Cyclist | 42.91 / 36.41 / 34.66 | 45.39 / 38.43 / 36.54 | 91.87 / 91.44 / 91.05 |

</details>

### Qualitative output

Projected 3D predictions from the primary Chen-split YOLO26s-3d checkpoint:

YOLO26s-3d projected KITTI predictions

<details>
<summary>KITTI R40 summary and training curves</summary>

YOLO26s-3d KITTI R40 summary

YOLO26s-3d training curves

</details>

Latency is batch-1 FP32 inference at `416x1280` on an RTX 4090 after warmup; preprocessing and postprocessing are excluded. Full class/difficulty R40 tables, raw benchmark JSON/CSV, checksums, checkpoints, ONNX exports, logs, and curves are in the [contributor artifact release](https://github.com/Dorablank/ultralytics/releases/tag/detect3d-v0.1.0).

The attached checkpoints were trained with an earlier contributor revision and revalidated/exported with the final PR code. They are research artifacts, not official Ultralytics weights.

## Validation

- `python -m ruff check .`
- `python -m ruff format --check .`
- 120 focused Detect3D, geometry, ONNX Runtime, plotting, and KITTI R40 tests passed on the latest upstream `main`.
- Final-code release check on both Chen n/s checkpoints:
  - PyTorch validation over 3,769 images with full KITTI R40
  - projected 3D PyTorch prediction
  - static opset-17 ONNX export with output shape `1x300x14`
  - ONNX checker and CUDAExecutionProvider prediction

## Data and licensing

KITTI data is not included in this PR or the artifact release. The README documents the expected layout, converted label schema, commands, checkpoint provenance, and KITTI terms. `THIRD_PARTY_NOTICES.md` attributes the MIT KITTI R40 protocol implementation used by the evaluator.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Adds native monocular 3D object detection support for KITTI datasets using dedicated YOLO11 and YOLO26 Detect3D models.

### 📊 Key Changes
- Introduces the `detect3d` task with KITTI dataset configuration, camera calibration handling, 3D label parsing, augmentation-aware projection matrices, and strict validation.
- Adds Detect3D model heads, training, validation, prediction, export, loss functions, geometry utilities, generic 3D metrics, and optional KITTI R40 evaluation.
- Extends results, plotting, serialization, ONNX export, and inference APIs to preserve and visualize 3D geometry.
- Adds extensive regression, loss, export, geometry, plotting, and KITTI evaluation tests, plus third-party licensing notices and README usage instructions.

### 🎯 Purpose & Impact
- Enables end-to-end monocular 3D detection workflows directly in Ultralytics, including training, validation, prediction, and ONNX deployment.
- Provides both generic 3D mAP metrics for model selection and opt-in KITTI fast/full evaluation for benchmark compatibility.
- Synchronizes image augmentations and camera calibration to improve geometric correctness, while retaining 2D supervision for filtered 3D targets.
- Expands the codebase substantially and introduces new geometry, evaluation, export, and dependency-integration paths that require thorough CI and compatibility review.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Dorablank, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your Detect3D work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for the new `detect3d` task, KITTI workflow, calibration requirements, and export behavior.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. The extensive Detect3D, geometry, export, plotting, and KITTI evaluation tests are a strong foundation.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). This is an automated welcome message; an engineer will assist with review and any follow-up questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean review. The Detect3D integration is consistently wired through task dispatch, dataset geometry handling, losses, validation, results, and export paths, with focused regression coverage for calibration transforms, metrics, KITTI evaluation, and runtime behavior. I found no evidence-backed issues requiring inline comments.


⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.

**Comment by Dorablank** [comment]
I have read the CLA Document and I sign the CLA

**Review by glenn-jocher — CHANGES_REQUESTED** [review]
This is not ready under the repository task contract and Core Principles:

- It adds `detect3d` to `TASKS`, but the canonical task order is explicitly `detect, segment, semantic, depth, classify, pose, obb`; introducing a separate task changes that public contract throughout the package.
- The implementation creates a parallel task stack (+6,475/-53 across 40 files) instead of extending the existing detection owner. This needs an owner-level redesign under `detect`, or an explicit project decision to change the canonical task architecture first; it cannot be repaired safely with incremental guards or compatibility scaffolding.

The branch is also currently conflicting with `main`, so validation on the recorded head is no longer sufficient.

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -560,6 +563,248 @@ def forward_head(
...
+
+    def postprocess(self, preds: torch.Tensor) -> torch.Tensor:
+        """Select one best class per end-to-end anchor and keep decoded 3D values aligned row by row.
+
+        Dimensions are decoded with the anchor's highest-scoring class prior in :meth:`_inference`. Allowing a secondary
+        class from the same anchor into the final top-k would therefore attach the wrong class prior to that row.
+        """
+        boxes, scores, d3_params = preds.split([4, self.nc, self.nd], dim=-1)
+        best_scores, best_class = scores.max(dim=-1, keepdim=True)
+        k = self.max_det if self.export else min(self.max_det, boxes.shape[1])
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/head.py:780]: 💡 **MEDIUM**: `self.export` forces `k=300` even when the feature maps contain fewer than 300 anchors. For valid small export sizes (for example `imgsz=32`, which yields only 21 P3–P5 anchors), `topk(300, ...)` raises an out-of-range error and prevents export. Clamp `k` to the available anchor count in export mode as well.

**Suggested change:**
```suggestion
        k = min(self.max_det, boxes.shape[1])
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The Detect3D integration is broadly well-covered, including calibration-aware transforms, loss paths, metrics, result serialization, and ONNX regression checks. One export edge case needs fixing: end-to-end Detect3D fails when the input produces fewer anchors than `max_det`.

💬 Posted 1 inline comment

⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.

**Comment by glenn-jocher** [comment]
Fresh UltralyticsAssistant review also reproduced an export defect on the current head: end-to-end Detect3D forces `topk(300)` even when a small input has fewer anchors (for example 21 anchors at `imgsz=32`), so export raises out of range. This reinforces the changes-requested decision. Please do not patch this in the parallel `detect3d` stack yet; the task must first be redesigned under the existing `detect` owner (or receive an explicit canonical-task architecture decision), then the shared detection postprocess should own the anchor clamp.

**Comment by Dorablank** [comment]
Thanks for the clarification. I understand that detect3d should not be introduced as a new canonical task, and I will not patch the current parallel task stack. My proposed redesign is to keep the canonical task as detect, infer monocular 3D capability from the model head or metadata, reuse the existing detection trainer, predictor, validator, and shared postprocess, and make calibration, 3D losses, KITTI evaluation, Results geometry, and export opt-in extensions of the detection owner. The anchor clamp would also be handled in the shared detection postprocess. Would this ownership model be acceptable? If so, would you prefer this PR to be reworked in place or replaced with smaller incremental PRs?

**Comment by glenn-jocher** [comment]
Yes, keeping `task=detect` with 3D capability owned by a specialized detection head and reusing shared detection infrastructure is the right direction. Please replace this PR with smaller incremental PRs, starting with the shared detection postprocess anchor clamp and focused regression tests, then introduce the minimal 3D head/loss path before calibration, metrics, `Results`, export, and docs in follow-ups.

**Comment by Dorablank** [comment]
Sounds good, thank you for clarifying the preferred architecture and sequence. I’ll start a fresh, focused PR from the current main branch for the shared detection postprocess anchor clamp and regression tests. Once the first replacement PR is up, I’ll mark this PR as superseded and link it here. After that lands, I’ll follow with the minimal 3D head/loss path and introduce the remaining pieces incrementally as outlined.

**Comment by Dorablank** [comment]
Superseded by #25771, the first incremental PR in the sequence above. It contains only the shared detection postprocess anchor clamp and focused regression coverage. I am closing this PR as planned; the minimal 3D head/loss path and the remaining integration pieces will follow in separate PRs after #25771 lands.
